core/src/error/mod.rs

```diff
@@ -371,13 +472,17 @@ impl Error {
         //
         // If span information is available, don't include the error property path
         // since it's redundant and not consistent with native compiler diagnostics.
-        match self.kind {
+        let diagnostic = match self.kind {
             ErrorKind::UnknownField(euf) => euf.into_diagnostic(self.span),
             _ => match self.span {
```

Comment:
I'd advise against using a discard match-arm here, because in case we end up adding a new variant that contains one or many diagnostics, they will get erased possibly without us noticing. Remove the discard here and the compiler will warn you instead.

Reply:
btw I don't understand the rationale behind `ErrorKind` being "hacked to be non-exhaustive" (as in, not using stdlib's `#[non_exhaustive]` attribute) ? And is it a good idea to be non-exhaustive, even ?

I think it's fair for `darling` to say "this is the set of errors we currently support", what am I missing ?

Reply:
This isn't a discard arm; we have special behavior for converting `ErrorUnknownField` into a `Diagnostic`, while for all the other `ErrorKind` variants we only need to use the `Display` impl of `ErrorKind` to get the diagnostic body.

> btw I don't understand the rationale behind ErrorKind being "hacked to be non-exhaustive" (as in, not using stdlib's #[non_exhaustive] attribute) ?

`darling` has an MSRV of 1.31; the `#[non_exhaustive]` attribute was [introduced in 1.40](https://blog.rust-lang.org/2019/12/19/Rust-1.40.0.html), so `darling` cannot use it.

> And is it a good idea to be non-exhaustive, even ?

It ends up not mattering at all, because `ErrorKind` is always going to remain internal to the `error` mod. The comments make it look like I was considering having `ErrorKind` be public at some point, but that was at least 4 years ago and I'm now 100% confident that `ErrorKind` will always stay internal. If you're interested in submitting a PR that removes the `__NonExhaustive` variant, I'd be happy to review one.

> I think it's fair for darling to say "this is the set of errors we currently support", what am I missing ?

It matters a lot that `darling` be able to add new error variants or _change the internal representation of error variants_ without any risk of breaking crate consumers or needing a minor version bump:
* Features like the "did you mean" suggestions would be much harder to implement if I had to worry about breaking code that tried to match on specific error types that `darling` supports.
* Adoption of features like "did you mean" would be slower if that was pushed as a minor version bump instead of a patch version bump.
* I think it's quite valuable that the output from using `Error::custom` is indistinguishable from the output of using other `Error` constructor functions, and that wouldn't be possible if people could interact with `ErrorKind`.